iree/compiler/Conversion/LinalgToSPIRV/VectorizeMemref.cpp

```diff
@@ -0,0 +1,307 @@
+//===- VectorizeMemref.cpp - Convert memref of scalar to vector------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Pass to convert memref into memref of vector.
+//
+//===----------------------------------------------------------------------===//
+#include "iree/compiler/Conversion/LinalgToSPIRV/Passes.h"
+#include "mlir/Dialect/StandardOps/IR/Ops.h"
+#include "mlir/Dialect/Vector/VectorOps.h"
+
+using namespace mlir;
+
+constexpr int vectorizationSizeInBits = 128;
+
+/// Return true of the type is a memref that can be vectorized to vector<4xi32>
+static bool memRefCanBeVectorized(Type type) {
+  auto memrefType = type.dyn_cast<MemRefType>();
+  if (!memrefType) return false;
+  if (memrefType.getElementType().isa<VectorType>()) return false;
+  if (vectorizationSizeInBits % memrefType.getElementTypeBitWidth())
+    return false;
+  if (ShapedType::isDynamic(memrefType.getShape().back())) return false;
+  if ((memrefType.getElementTypeBitWidth() * memrefType.getShape().back()) %
+          vectorizationSizeInBits !=
+      0)
+    return false;
+  return true;
+}
+
+/// Return the bitwith of a scalar or vector type.
+static unsigned getBitWidth(Type type) {
```

Comment:
Do we want to just have this in MLIR?

Also instead of asserting, make the return type `Optional<unsigned>` and return `{}` on failure.